src/miner.cpp

```diff
@@ -717,23 +771,38 @@ std::optional<MinerAddress> ExtractMinerAddress::operator()(const libzcash::Sapl
     return addr;
 }
 std::optional<MinerAddress> ExtractMinerAddress::operator()(const libzcash::UnifiedAddress &addr) const {
-    for (const auto& receiver: addr) {
-        if (std::holds_alternative<libzcash::SaplingPaymentAddress>(receiver)) {
-            return std::get<libzcash::SaplingPaymentAddress>(receiver);
-        }
+    auto preferred = addr.GetPreferredRecipientAddress(consensus, height);
+    if (preferred.has_value()) {
+        std::optional<MinerAddress> ret;
+        std::visit(match {
+            [&](const libzcash::OrchardRawAddress addr) { ret = MinerAddress(addr); },
+            [&](const libzcash::SaplingPaymentAddress addr) { ret = MinerAddress(addr); },
+            [&](const CKeyID keyID) { ret = operator()(keyID); },
+            [&](const auto other) { ret = std::nullopt; }
+        }, preferred.value());
+        return ret;
+    } else {
+        return std::nullopt;
     }
-    return std::nullopt;
 }
 
 
-void GetMinerAddress(MinerAddress &minerAddress)
+void GetMinerAddress(std::optional<MinerAddress> &minerAddress)
 {
     KeyIO keyIO(Params());
 
+    // If the user sets a UA miner address with an Orchard component, we want to ensure we
+    // start using it once we reach that height.
+    int height;
+    {
+        LOCK(cs_main);
+        height = chainActive.Height() + 1;
+    }
+
     auto mAddrArg = GetArg("-mineraddress", "");
     auto zaddr0 = keyIO.DecodePaymentAddress(mAddrArg);
     if (zaddr0.has_value()) {
-        auto zaddr = std::visit(ExtractMinerAddress(), zaddr0.value());
+        auto zaddr = std::visit(ExtractMinerAddress(Params().GetConsensus(), height), zaddr0.value());
```

Comment:
I'd propagate this parameter outward to make it an explicit argument of `GetMinerAddress`.